src/ProjectOrigin.VerifiableEventStore/Services/Repository/InMemory/InMemoryRepository.cs

```diff
@@ -69,29 +73,60 @@ public Task FinalizeBlock(BlockHash hash, ImmutableLog.V1.BlockPublication publi
         }
     }
 
-    public Task<ImmutableLog.V1.Block?> GetBlock(TransactionHash transactionHash)
+    public Task<Registry.V1.Block?> GetBlock(TransactionHash transactionHash)
     {
         lock (_lockObject)
         {
             var e = _events.SingleOrDefault(x => x.TransactionHash == transactionHash);
 
             if (e is null)
-                return Task.FromResult<ImmutableLog.V1.Block?>(null);
+                return Task.FromResult<Registry.V1.Block?>(null);
 
             var index = _events.IndexOf(e);
 
             var block = _blocks.Values.SingleOrDefault(x => x.FromTransaction <= index && index <= x.ToTransaction);
 
             if (block is null)
-                return Task.FromResult<ImmutableLog.V1.Block?>(null);
+                return Task.FromResult<Registry.V1.Block?>(null);
 
-            var result = new ImmutableLog.V1.Block
+            var result = new Registry.V1.Block
             {
                 Header = block.Header,
                 Publication = block.Publication
             };
 
-            return Task.FromResult<ImmutableLog.V1.Block?>(result);
+            return Task.FromResult<Registry.V1.Block?>(result);
+        }
+    }
+
+    public Task<IList<Block>> GetBlocks(int skip, int take, bool includeTransactions)
+    {
+        lock (_lockObject)
+        {
+            var data = _blockList.Skip(skip).Take(take).Select(x =>
+            {
+                var block = new Block
+                {
+                    Header = x.Header,
+                    Publication = x.Publication,
+                    Height = _blockList.IndexOf(x) + 1
+                };
+
+                if (includeTransactions)
+                {
+                    var transactions = _events
+                        .Skip(x.FromTransaction)
+                        .Take(x.ToTransaction - x.FromTransaction + 1)
+                        .Select(y => Transaction.Parser.ParseFrom(y.Payload))
+                        .ToList();
+                    block.Transactions.AddRange(transactions);
+                }
+
+                return block;
+            }
+            ).ToList();
+
+            return Task.FromResult<IList<Block>>(data);
```

Comment:
protecting shared resources is correct, but wrapping this logic inside an asynchronous method that returns Task.FromResult is confusing to read for me. I understand it as the method not actually being truly asynchronous. You are going back to making threads wait. Consider SemaphoreSlim with WaitAsync. Så kan applikationen "awaite" låsen, asynkront.